### PR TITLE
scheduler(ticdc): split span by region written keys

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: ref #7720

### What is changed and how it works?

Support split span by region written keys by evenly distribute write throughput to every TiCDC nodes.

#### Tests 

 - Unit test

##### Will it cause performance regression or break compatibility?

No.

##### Do you need to update user documentation, design documentation or monitoring documentation?

No.

### Release note 

```release-note
Support split span by region written keys by evenly distribute write throughput to every TiCDC nodes.
```

### CODE REVIEW

File: cdc/scheduler/internal/v3/keyspan/splitter_region_count.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	bo := tikv.NewBackoffer(ctx, 500)
+	regions, err := m.regionCache.ListRegionIDsInKeyRange(bo, span.StartKey, span.EndKey)
+	if err != nil {
+		log.Warn("schedulerv3: list regions failed, skip split span",
+			zap.String("namespace", m.changefeedID.Namespace),
+			zap.String("changefeed", m.changefeedID.ID),
+			zap.Error(err))
+		return []tablepb.Span{span}
+	}
+	if len(regions) <= config.RegionThreshold || totalCaptures == 0 {
```
Comment (nongfushanquan) [inline cdc/scheduler/internal/v3/keyspan/splitter_region_count.go:56]: totalCaptures <=1 ?

File: cdc/scheduler/internal/v3/keyspan/splitter_write.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	pageStartIdx, pageLastIdx := 0, 0
+	for i := 1; i < pages; i++ {
+		for idx := pageStartIdx; idx < len(regions); idx++ {
+			restPages := pages - i
+			restRegions := len(regions) - idx
+			pageLastIdx = idx
+			currentWrittenKeys := regions[idx].WrittenKeys
+			if (idx > pageStartIdx) &&
+				((restPages >= restRegions) ||
+					(2*accWrittenKeys+currentWrittenKeys > 2*uint64(i)*writtenKeysPerPage)) {
```
Comment (nongfushanquan) [inline cdc/scheduler/internal/v3/keyspan/splitter_write.go:114]: why the number 2 is needed?

File: cdc/scheduler/internal/v3/keyspan/reconciler.go [hunk 3]
```diff
@@ -108,7 +117,13 @@ func (m *Reconciler) Reconcile(
 			tableSpan := spanz.TableIDToComparableSpan(tableID)
 			spans := []tablepb.Span{tableSpan}
 			if compat.CheckSpanReplicationEnabled() {
-				spans = m.splitter.split(ctx, tableSpan, len(aliveCaptures), m.config)
+				for _, splitter := range m.splitter {
+					spans = splitter.split(ctx, tableSpan, len(aliveCaptures), m.config)
+					if len(spans) <= 1 {
+						continue
```
Comment (3AceShowHand) [inline cdc/scheduler/internal/v3/keyspan/reconciler.go:123]: revert the condition, only break once.

File: pkg/pdutil/api_client_test.go [hunk 4]
```diff
@@ -162,3 +165,95 @@ func TestMetaLabelDecodeJSON(t *testing.T) {
...
+	}
+	rs, err := pc.scanRegions(context.Background(), tablepb.Span{}, []string{mockPDServer.URL}, 1)
+	require.Nil(t, err)
+	require.Equal(t, 7, len(rs))
+
+	handler = func(startKey, endKey []byte, limit int) RegionsInfo {
+		return RegionsInfo{Regions: regions}
+	}
+	rs, err = pc.scanRegions(context.Background(), tablepb.Span{}, []string{mockPDServer.URL}, 1024)
+	require.Nil(t, err)
```
Comment (3AceShowHand) [inline pkg/pdutil/api_client_test.go:219]: for all errors, use `require.NoError`

**Review by 3AceShowHand — APPROVED** [review]

File: pkg/pdutil/api_client.go [hunk 5]
```diff
@@ -132,6 +151,129 @@ func (pc *pdAPIClient) UpdateMetaLabel(ctx context.Context) error {
...
+		ID:          regionID,
+		StartKey:    hex.EncodeToString(start),
+		EndKey:      hex.EncodeToString(end),
+		WrittenKeys: writtenKeys,
+	}
+}
+
+// RegionInfo records detail region info for api usage.
+// NOTE: This type is a copy of github.com/tikv/pd/server/api.RegionInfo.
+// To reduce dependency tree, we do not import the api package driectly.
```
Comment (zhaoxinyu) [inline pkg/pdutil/api_client.go:166]:
```suggestion
// To reduce dependency tree, we do not import the api package directly.
```

File: pkg/pdutil/api_client.go [hunk 6]
```diff
@@ -132,6 +151,129 @@ func (pc *pdAPIClient) UpdateMetaLabel(ctx context.Context) error {
...
+type RegionInfo struct {
+	ID       uint64 `json:"id"`
+	StartKey string `json:"start_key"`
+	EndKey   string `json:"end_key"`
+
```
Comment (zhaoxinyu) [inline pkg/pdutil/api_client.go:171]:
```suggestion
```

**Review by zhaoxinyu — APPROVED** [review]

**Pull request merged**